components/client/src/main/java/com/hotels/styx/client/StyxBackendServiceClient.java

```diff
@@ -157,13 +159,29 @@ private static boolean isHeadRequest(LiveHttpRequest request) {
                     .onErrorResumeNext(cause -> {
                         RetryPolicyContext retryContext = new RetryPolicyContext(this.id, attempt + 1, cause, request, previousOrigins);
                         return retry(request, retryContext, newPreviousOrigins, attempt + 1, cause);
-                    });
+                    })
+                    .map(response -> addOriginId(host.id(), response))
+                    .onErrorResumeNext(this::mapExceptionCauses);
         } else {
             RetryPolicyContext retryContext = new RetryPolicyContext(this.id, attempt + 1, null, request, previousOrigins);
             return retry(request, retryContext, previousOrigins, attempt + 1, new NoAvailableHostsException(this.id));
         }
     }
 
+    private Observable<LiveHttpResponse> mapExceptionCauses(Throwable cause) {
+        if (cause instanceof NoAvailableHostsException) {
+            return Observable.error(new NoAvailableHostsException(id));
```

Comment:
The stack trace seems bit pointless for this exception because we know where and why the exception is originally thrown.

However this comment highlights an interesting issue: The `NoAvailableHostsException` is thrown from the `Transport` class when it doesn't have a connection pool configured. But then again it doesn't make sense to create a `Transport` object without pool. Instead, the exception should be thrown earlier. Before the `Transport` is even instantiated.

Reply:
Ok. This is effectively dead code. It used to be the case that `Transport` checked if the chosen origin/pool was available or not. But since refactoring `StyxBackendServiceClient` it is no longer possible to use transport/send without providing a connection pool.

So this exception mapping can be removed.